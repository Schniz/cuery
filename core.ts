type Primitive = string | boolean | number;
export type Param<T> = (data: T) => any;
export type AstNode<InputData> =
  | {
      type: "computed";
      value: Param<InputData>;
    }
  | {
      type: "primitive";
      value: Primitive;
    }
  | {
      type: "string";
      value: string;
    };

export abstract class Query<Input, Output> {
  nodes: AstNode<Input>[];
  constructor(nodes: AstNode<Input>[]) {
    this.nodes = nodes;
  }
}

export type SqlFunctionParam<Input, Output> =
  | AstNode<Input>[]
  | Query<Input, Output>
  | Param<Input>
  | Primitive;

export function parse<Input, Output>(
  strings: TemplateStringsArray,
  ...params: SqlFunctionParam<Input, Output>[]
): AstNode<Input>[] {
  const nodes = [] as AstNode<Input>[];

  for (const index in strings) {
    const string = strings[index];
    const param = params[index];

    nodes.push({ type: "string", value: string });

    if (!param) {
    } else if (param instanceof Query) {
      nodes.push(...param.nodes);
    } else if (Array.isArray(param)) {
      nodes.push(...param);
    } else if (typeof param === "function") {
      nodes.push({ type: "computed", value: param });
    } else {
      nodes.push({ type: "primitive", value: param });
    }
  }

  return nodes;
}
