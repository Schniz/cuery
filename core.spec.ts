import { parse } from "./core";

test("parses a query", () => {
  const nodes = parse<{ age: number }, {}>`
    SELECT first_name, last_name, age FROM table
    WHERE first_name = ${"Gal"}
    AND age = ${x => x.age}
  `;

  expect(nodes).toMatchSnapshot();
});

test("composes correctly", () => {
  const allUsers = parse<{}, {}>`
    SELECT * FROM users
  `;
  const nodes = parse<{}, {}>`
    SELECT * FROM (${allUsers}) all_users;
  `;
  expect(nodes).toMatchSnapshot();
});
