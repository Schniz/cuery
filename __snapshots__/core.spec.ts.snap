// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composes correctly 1`] = `
Array [
  Object {
    "type": "string",
    "value": "
    SELECT * FROM (",
  },
  Object {
    "type": "string",
    "value": "
    SELECT * FROM users
  ",
  },
  Object {
    "type": "string",
    "value": ") all_users;
  ",
  },
]
`;

exports[`parses a query 1`] = `
Array [
  Object {
    "type": "string",
    "value": "
    SELECT first_name, last_name, age FROM table
    WHERE first_name = ",
  },
  Object {
    "type": "primitive",
    "value": "Gal",
  },
  Object {
    "type": "string",
    "value": "
    AND age = ",
  },
  Object {
    "type": "computed",
    "value": [Function],
  },
  Object {
    "type": "string",
    "value": "
  ",
  },
]
`;
